force-app/test/package/classes/unit/ArgumentTest.cls

```diff
@@ -75,29 +75,32 @@ public class ArgumentTest {
   }
 
   @isTest
-  static void givenEqualsMatchable_matchesCustomType() {
+  static void givenEqualsMatchable_matchesCustomTypeWithEquals() {
     // Arrange
     Argument.Matchable sut = Argument.equals(new CustomTypeWithEquals('test', 10, new Account(Name = 'test')));
 
     // Act & Assert
     Assert.isTrue(sut.matches(new CustomTypeWithEquals('test', 10, new Account(Name = 'test'))));
-
+    
+    // TODO: Fix commented out asserts; they throw a GACK if uncommented.
```

Comment:
@scolladon I am stuck on this. It seems to impact only Boolean and Integer inputs. Since it's a GACK, I have definitely been struggling on how to best diagnose this...my only lead is that it appears to happen while evaluating the Object inside of `CustomTypeWithEquals.equals()`.

Reply:
I reproduce the issue...
I tried to downgrade the API but it produces the same result...
When I use `==` it works...

Then I propose to change the implementation of `EqualsMatchable.matches` to be :
```apex
 public Boolean matches(final Object callArgument) {
      if (this.callArgumentToMatch == null) {
        return callArgument == null;
      }

      // case sensitive check for string
      if (this.callArgumentToMatch instanceof String) {
        return this.callArgumentToMatch.equals(callArgument);
      }

      // native platform equality check else
      return this.callArgumentToMatch == callArgument;
    }
```